### ISSUE TITLE
[Question] Best way to manage variables in a project with many ink files?

### ISSUE THREAD

**Issue by Ludipe** [inkle/ink#109]
Hey! We're working on quite a big project using Ink (and Unity), so we are trying to run as many tests as possible to figure out best practices, trying to avoid things that may be an inconvenience in a few months from now. 

Ours is a game where you move a character around, when you press the "speak button" close to another character then ink takes control and triggers the dialogue interface. 

There are many things we want to keep track of. From money or reputation with a certain character to "did I ask him X". We want to be able to access and modify those values even if we're not on a dialogue (as well as efficiently save them to manage save slots), for example, knocking on somebody else door (outside of dialogue)  might open a new dialogue branch when talking to a certain character.

We're still debating what would be the best way to structure everything, as we're still getting used to working with ink. Should we use include statements to every ink file in an "ink master file" and just save that one? create variables in Unity to keep track of all the values that should be available everywhere? Just save the state of every dialogue?

It may be a silly question, but we're trying to set a solid base to build the game and we weren't really sure about the best solution for this.

**Comment by joethephish** [comment]
Yeah, deciding this structure is surprisingly non-trivial, and we've had similar discussions on each of our games.

Firstly, one of your questions is easy: You should definitely have a master ink file, and include other "scene" ink files from that one. Ink is designed to have a single Story object which you interact with, so it can store global state.

In terms of deciding where the "source of truth" is stored for certain variables: We like to store story-relevant variable within the story as ink variables, and then hook those into the game side. One useful side effect is that the writer/designer can trivially create new variables in the story, and you don't need to remember to serialise them to the save state, since once you've saved the overall story state, the rest is automatic.

In 80 Days, we didn't link the map locations together using ink - we had a game data file where we defined which knot a city/journey jumps to. However, you could imagine a game in which it's more convenient to include location choices as ink choices.

Hope that helps!

**Comment by lhughes41** [comment]
Should ink variables declared via 'VAR' in *INCLUDED* ink files be visible inside the master ink file (where INCLUDE was declared).
 I am asking from a programmatic context. My goal is to keep a bunch of vars declared in a single file that is imported via INCLUDE for a series of different master files.
  Programmatically my problem is when I try to assign to an "INCLUDEd" var in such a master story instance via 'variablesState.set_Item' I get back:
  "Ink.Runtime.StoryException: Cannot assign to a variable (is_this_a_campaign) that hasn't been declared in the story
   at Ink.Runtime.VariablesState.set_Item(String variableName, Object value)"

I get no such complaints from set_Item or Get_item for that matter when I ask for values or set values for variables set inside the master ink file.

Summary: can VARs declared in an "INCLUDE"d file be used in a master file?

thank you!

**Comment by Selsynn** [comment]
Yes; you can have the variable declared in the included files, BUT you have to always open or compile, or whatever the MASTER file.

You can have 
Master file (containing the include line)
V file => where you declare your variable
U file => where you use your variable declared

When opening in inky the Master file, you can edit U file or V file or whatever you want, and just be happy where you declared your variable and where you use it. 

However, if you open U directly, without going from Master, so without the link with the V file, it won't find the file, and won't know any variable.

**Comment by lhughes41** [comment]
Selsynn my you are quick! I tried to delete my false alarm post like 30 minutes after I did it!  But you spotted it ;-) Kudos to you and sorry to waste your time.

 I'm the programming side and so don't write much ink so I wrote '#INCLUDE' in my ink rather than "INCLUDE"  :-( So it was all my bad.

   I wrote here rather than a new thread because I was thinking it was good practice to continue under a relevant thread and the thread didn't seem to be 'closed' (in fact still doesn't).

And again thanks for quick thoughtful reply.
